lib/iris/analysis/_area_weighted.py

```diff
@@ -1150,3 +459,266 @@ def regrid_callback(*args, **kwargs):
+    # This is likely to become switchable in the future, see:
+    # https://github.com/SciTools/iris/issues/5461
+    oob_invalid = True
+
+    data_shape = data.shape
+    if ma.is_masked(data):
+        unmasked = ~ma.getmaskarray(data)
+        # Calculate contribution from unmasked sources to each target point.
+        weight_sums = _standard_regrid_no_masks(unmasked, weights, tgt_shape)
+    else:
+        # If there are no masked points then all contributions will be
+        # from unmasked sources, so we can skip this calculation
+        weight_sums = np.ones(data_shape[:-2] + tgt_shape)
+    mdtol = max(mdtol, 1e-8)
+    tgt_mask = weight_sums > 1 - mdtol
+    # If out of bounds sources are treated the same as masked sources this
+    # will already have been calculated above, so we can skip this calculation.
+    if oob_invalid or not ma.is_masked(data):
+        # Calculate the proportion of each target cell which is covered by the
+        # source. For the sake of efficiency, this is calculated for a 2D slice
+        # which is then broadcast.
+        inbound_sums = _standard_regrid_no_masks(
+            np.ones(data_shape[-2:]), weights, tgt_shape
+        )
+        if oob_invalid:
+            # Legacy behaviour, if the full area of a target cell does not lie
+            # in bounds it will be masked.
+            oob_mask = inbound_sums > 1 - 1e-8
+        else:
+            oob_mask = inbound_sums > 1 - mdtol
+        # Broadcast the mask to the shape of the full array
+        oob_slice = ((np.newaxis,) * len(data.shape[:-2])) + np.s_[:, :]
+        tgt_mask = tgt_mask * oob_mask[oob_slice]
+
+    # Calculate normalisations.
+    normalisations = np.ones_like(weight_sums)
+    normalisations *= tgt_mask
+    normalisations[tgt_mask] /= weight_sums[tgt_mask]
+
+    # Mask points in the result.
+    if ma.isMaskedArray(data):
+        # If the source is masked, the result should have a similar mask.
+        fill_value = data.fill_value
+        normalisations = ma.array(
+            normalisations, mask=~tgt_mask, fill_value=fill_value
+        )
+    elif np.any(~tgt_mask):
+        normalisations = ma.array(normalisations, mask=~tgt_mask)
+
+    # Use input cube dtype or convert values to the smallest possible float
+    # dtype when necessary.
+    dtype = np.promote_types(data.dtype, np.float16)
+
+    # Perform regridding on unmasked data.
+    result = _standard_regrid_no_masks(
+        ma.filled(data, 0.0), weights, tgt_shape
+    )
+    # Apply normalisations and masks to the regridded data.
+    result = result * normalisations
+    result = result.astype(dtype)
+    return result
+
+
+def _regrid_along_dims(data, x_dim, y_dim, weights, tgt_shape, mdtol):
+    """Regrid data, handling masks and dimensions."""
+    # Handle scalar coordinates.
+    # Note: scalar source coordinates are only handled when their
+    # corresponding target coordinate is also scalar.
+    if x_dim is None:
+        x_none = True
```

Comment:
Can simplify this code by removing `x_none` and `y_none`, and instead just tracking the number of added scalar coords. This also effects the `np.squeeze` at the bottom of this code.